### ISSUE TITLE
Provide org.kde.kwalletd5 dbus interface

### ISSUE THREAD

**Issue by cryptomilk** [keepassxreboot/keepassxc#3679]
KDE uses the kwallet password manager to provide a DBUS API for password storage. For this they don't use the `org.freedesktop.secrets` API but have provide their own via `org.kde.kwalletd5`.

There have been efforts to move to `org.freedesktop.secrets` since 2013 but nothing has happened yet [1].

Till KDE supports `org.freedesktop.secrets` it would be great if keepassxc would provide the DBUS API of kwallet.

## Expected Behavior
Provide `org.kde.kwalletd5` DBUS API

```
createFolder
folderList
entryList
writeEntry
writePassword
readPassword
```

## Current Behavior
Doesn't provide `org.kde.kwalletd5` DBUS API

[1] https://bugs.kde.org/show_bug.cgi?id=313216

**Comment by droidmonkey** [comment]
I'm not keen on supporting a one-off "standard" that needs to be replaced by the actual standard.

**Comment by cryptomilk** [comment]
FYI https://marc.info/?l=kde-devel&m=146788782429932&w=2

**Comment by Aetf** [comment]
I'm on the KDE desktop and I've been using KeepassXC as the libsecret provider for a while (since I implemented it :P).

KDE applications are luckily flexible enough that many of them can be configured to use libsecret-based API either via settings or via some environment variable which they use to detect the desktop environment. This includes chromium-based browsers and many electron apps.

There are still a few things before I can disable kwallet system-wide:

* KCM `Online Accounts`, this saves google accounts credentials if you want to use `kio-gdrive`
* Network manager, when wireless passwords are saved for the user only

I'm now playing around with [signon](https://accounts-sso.gitlab.io/), which is what KCM `Online Accounts` uses for secret saving, which can be configured to use libsecret backend. I'll update once I feel it's good for everyday use.

For network manager, I'm sure there are also some settings/envvar to tweak but I haven't look into it yet.

Anyway, the bottom line is that libsecret DBus interface can still be helpful for KDE users.

**Comment by cryptomilk** [comment]
I'm not saying it is not useful, but most apps only use kwallet like akonadi or the plasma network manager applet. It will take while till KDE fully supports libsecret.

**Comment by Aetf** [comment]
Yeah, I just feel this is kind of a chicken-egg problem. Kwallet works so well for KDE applications and there is little incentive for app/framework developers to push for the standard. I just try to start using libsecret, at least for myself, hoping it will happen a little sooner ;)

**Comment by droidmonkey** [comment]
Aetf what would it take to support this API if you could somehow detect that it was needed either on the fly or via explicit setting?

**Comment by Aetf** [comment]
From a quick glance of kwallet API, the high-level concepts look similar.  Implementation-wise, some shim code is needed as the method semantics are probably different.

Detection on the fly doesn't make sense because it's the other way around: the client checks if any service is available at the predefined DBus address. But it's possible to register both DBus addresses (kwallet and libsecret) at the same time and expose methods correspondingly. This can be covered by a setting.

So, in theory, one could have KeepassXC acting as both a libsecret provider and a kwallet provider. But don't expect passwords set by different APIs to be compatible. Maybe better to use to separate groups for each API.

**Comment by droidmonkey** [comment]
Yah that is too messy for me... if you think of a clean integration we can consider it, but I don't think it is worth the effort.

**Comment by azrdev** [comment]
Building the sim standalone sounds more feasible, i.e. an extra application providing the kwallet API but only proxying to the libsecret one.

**Comment by 0xpr03** [comment]
Just to leave a plasma-user voice: I just use keepassxc as the main manager, if an KDE applications requires credentials I'll provide them via autotype/copy 'n paste, that's it. Even more I'm quite okay with kwallet getting used for WiFi as I don't see why I need to unlock my complete keepass just for my WiFi to run after logging in. At least to me it's a huge difference between unlocking my WiFi and unlocking my global password manager.

**Comment by NiklasBeierl** [comment]
@Aetf can you provide some details as to how you make chrome / other applications run using KPXC as "libsecret provider"?
Might be a noob question, but is the above equivalent to making an application use the "org.freedesktop.secrets" DBUS api? 

I personally would like to use KPXC as the credentials store for:
- Nextcloud
- Chrome
- plasma-nm / NetworkManager
- Pycharm

WRT to [the tragedy that seems to be kwalletd5](https://bugs.kde.org/show_bug.cgi?id=313216):

I do not know how exactly applications determine what secrets provider to use, but would it be possible to add a "system setting" to KDE, which somehow lets applications know whether they should use "org.freedesktop.secrets" or "org.kde.kwalletd5"? Or do applications just look at the DE they are running in and then determine to go for KWallet / the free desktop one?

**Comment by soredake** [comment]
https://invent.kde.org/frameworks/kwallet/-/merge_requests/11

**Comment by Aetf** [comment]
>  is the above equivalent to making an application use the "org.freedesktop.secrets" DBUS api?

Yes. `libsecret` is the client library that talks to a secret service server using the "org.freedesktop.secrets" DBus API. And I usually use "secret service API" to refer to the thing.

> I do not know how exactly applications determine what secrets provider to use, but would it be possible to add a "system setting" to KDE, which somehow lets applications know whether they should use "org.freedesktop.secrets" or "org.kde.kwalletd5"? Or do applications just look at the DE they are running in and then determine to go for KWallet / the free desktop one?

Unfortunately, there's no universal way to do this. It depends on the application itself or the keyring library it uses to determine which secret API to use. Usually, this involves looking at several environment variables to determine the DE and try several APIs and pick whichever one is first available.

So at the very least, maybe you can forge the environment variable to make it appear like gnome when starting the app, but that has other consequences as password storage isn't the only thing that changes its behavior depending on the DE.

There are some supported override methods specific to the application and library, though. For example,

* python [keyring](https://pypi.org/project/keyring/) based applications can use a config file to force select a backend: https://github.com/jaraco/keyring#configuring
* almost all chromium based browsers accept a command-line argument `--password-store=gnome` to force the use of "gnome" keyring, which is actually the secret service API.
* I only have experience for CLion, but I guess it's the same across JetBrain IDEs: when you choose the native keychain option, it will prefer secret service API when both that and KDE wallet is present: https://www.jetbrains.com/help/clion/reference-ide-settings-password-safe.html

I don't know about other apps. But I'm sure there's a similar way to select the backend.

Personally, I haven't considered switch plasma-nm/NetworkManager to secret service because it starts up early after login, and I haven't figured out an elegant and clean way to start and unlock KeePassXC yet. (This has some hope given the transition to using systemd to handle session startup in Plasma, and I have some idea about auto unlocking using PAM, but I haven't tried).

And there's the kde-accounts KCM module, which I think provides account info for kio clients (like kio-gdrive and samba), and I think there's no easy to switch these to secret service API.

> https://invent.kde.org/frameworks/kwallet/-/merge_requests/11

I'd appreciate it if they decide to switch KDE apps to use qtkeyring, that will make the switching much easier.

**Comment by sunjam** [comment]
>     https://invent.kde.org/frameworks/kwallet/-/merge_requests/11

This was released in KDE Framework 5.97 on August 14th.  [See changelog](https://kde.org/announcements/frameworks/5/5.97.0/)
"Kwallet introduces Secret Service API".

**Comment by michaelk83** [comment]
> "Kwallet introduces Secret Service API".

This rather the reverse of what was asked here, isn't it? It's a Secret Service API frontend for KWallet backend storage, while this issue asks for a KWallet API frontend for Secret Service backend storage. But hopefully this will encourage more KDE apps to switch over to the Secret Service API, either via `libsecret` or preferably via QtKeyChain. In the long-run, it should be easier to switch the two backends more easily. So still good news.

**Comment by sunjam** [comment]
Would one of you be kind enough to open an issue on [the KWallet repo](https://invent.kde.org/frameworks/kwallet/-/issues).  @michaelk83 Sounds like you have a good understanding of this request and could explain it to them + drop them a link back to this github issue.

**Comment by michaelk83** [comment]
I don't think it makes sense for KWallet or KDE devs to implement this. Apps that use the KWallet API can already use KWallet as is. KDE devs' time would be better spent porting more apps to QtKeyChain.

I posted a request to deprecate the KWallet API to encourage that transition:
https://bugs.kde.org/show_bug.cgi?id=458318

**Comment by ThisNekoGuy** [comment]
[Instead, we should check for QtKeyChain,](https://bugs.kde.org/show_bug.cgi?id=459039#c2) I was discussing this over on bugs.kde.org about how Kwallet should really be replaced with a KDE-maintained fork of KeePassXC for better Plasma integration and it seems to be a welcomed thought over there since Kwallet doesn't get much work and is objectively inferior at achieving what KeePassXC does.

That said, I was advised to make this QtKeyChain (or libsecret) suggestion in the meantime so that KeePassXC can have better intergration with `kwalletd5` until something gets sorted out.

**Comment by droidmonkey** [comment]
> KDE-maintained fork of KeePassXC

Omg no please...

**Comment by 0xpr03** [comment]
> KDE-maintained fork of KeePassXC

Please don't do that. You'll just be creating a lot of weird user issues and confusion.

Or what else do you intend to do in an "official fork"? It just sounds like something that will get behind in feature parity over time, getting mixed up with KPXC and people going back to installing the original.

Maybe rather try to land your utmost requirements in keepassxc and just ship that, like unchanged ? I dropped kwallet the moment I installed keepassXC, it's now delegated to providing wifi passwords towards the OS.

~Written vom KDE Plasma 5.

**Comment by michaelk83** [comment]
> Instead, we should check for QtKeyChain

Just to clarify, QtKeyChain is only relevant if KeePassXC ever decides to act as a *client* for KWallet or other Secret Service providers, displaying and managing secrets from those sources (along with it's own), or to import secrets from those sources.

This current issue (3679) is about KeePassXC acting as a *replacement* for KWallet et al. It can already do so for client apps that support Secret Service, but not for those that only know the old KWallet API (hence this issue). It is *those* apps that should be migrating to QtKeyChain, not KeePassXC.

Regarding a possible KDE fork, how I imagined it might work, was that the devs here would continue as before, and KDE devs would only add their own integration patches downstream, to provide a more seamless experience with KDE. But that is up to KDE devs. I agree that this can cause some user confusion.

**Comment by ThisNekoGuy** [comment]
> Regarding a possible KDE fork, how I imagined it might work, was that the devs here would continue as before, and KDE devs would only add their own integration patches downstream, to provide a more seamless experience with KDE. But that is up to KDE devs. I agree that this can cause some user confusion.

How I imagined it was that the fork name could also be changed to avoid confusion with KeePassXC; the interface might look alike but there would be a clear distinction between what is and isn't "officially" KeePassXC. The fork could even refer users that desire to sponsor the project here; it doesn't have to be confusing, but it can be if the execution's not properly thought through.

**Comment by droidmonkey** [comment]
That is a terrible idea and will absolutely generate confusion and distrust in the product. Just write the code and merge it into our program. That is exactly what was done with the secret service api.

**Comment by rejedai** [comment]
Are there any updates in that direction? I'm currently experiencing that electron based applications by default try to use with kwalletd even though it is disabled, and don't even try to use freedesktop secrets...

secret-tool worked without any problems.

**Comment by ThisNekoGuy** [comment]
@rejedai My solution to the Kwallet problem is to kill the Kwallet process on login, because it would run anyway for some reason:
```bash
[Desktop Entry]
Type=Application
Name=Kwalletd5 Killer
Comment=Stop Kwallet from launching as the secrets manager
TryExec=sh
Exec=sh -c "kill (pidof kwalletd5)"
Icon=kwalletmanager
StartupNotify=false
NoDisplay=true

```

**Comment by rejedai** [comment]
I don't have any instances running. According to the log the program can't get the status of kwalletd and tries to run it, when it fails it just stops using wallet. 

```
[12372:1221/143056.850927:ERROR:object_proxy.cc(590)] Failed to call method: org.kde.KWallet.isEnabled: object_path= /modules/kwalletd6: org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
[12372:1221/143056.850941:ERROR:kwallet_dbus.cc(107)] Error contacting kwalletd6 (isEnabled)
[12372:1221/143056.851236:ERROR:object_proxy.cc(590)] Failed to call method: org.kde.KLauncher.start_service_by_desktop_name: object_path= /KLauncher: org.freedesktop.DBus.Error.ServiceUnknown: The name org.kde.klauncher was not provided by any .service files
[12372:1221/143056.851244:ERROR:kwallet_dbus.cc(79)] Error contacting klauncher to start kwalletd6
[12372:1221/143056.972455:ERROR:object_proxy.cc(590)] Failed to call method: org.kde.KWallet.close: object_path= /modules/kwalletd6: org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
[12372:1221/143056.972469:ERROR:kwallet_dbus.cc(459)] Error contacting kwalletd6 (close)
```

**Comment by pauschuu** [comment]
> I don't have any instances running. According to the log the program can't get the status of kwalletd and tries to run it, when it fails it just stops using wallet.

### tl;dr:
Chromium/Chrome has a command line option to force the usage of libsecret.
```sh
google-chrome-stable --password-store=gnome-libsecret
```

### Full Post
I tried changing environment variables until I found that
```sh
export XDG_CURRENT_DESKTOP=invalidfoobar
unset KDE_FULL_SESSION
```
will stop Chrome from trying to use kwallet. But it still doesn't use libsecret in this case though.

Fortunately, the Arch Wiki comes to the rescue (as always):
https://wiki.archlinux.org/title/chromium#Force_a_password_store
```sh
google-chrome-stable --user-data-dir=/tmp/chrome-test-profile --password-store=gnome-libsecret
```
will force the usage of the Secret Service API.
(You don't even need the env modifications)

I know it's just one application (I guess this flag isn't working across all Electron Apps [or is it?])

**Comment by Lgmrszd** [comment]
Instead of providing kwallet interface, apps should be using secretservice interface, as it was suggested earlier. For me, the only piece of software still not supporting secretservice is the plasma-nm (NetworkManager applet). Seems like it has [issue tracker available](https://invent.kde.org/plasma/plasma-nm/-/issues/), but it requires KDE Identity account available. I'd appreciate if someone with an account could open an issue here!

**Comment by cryptomilk** [comment]
Done: https://invent.kde.org/plasma/plasma-nm/-/issues/7

**Comment by ap-1** [comment]
Here is what I've understood from skimming this thread:

[484045](https://bugs.kde.org/show_bug.cgi?id=484045) was opened on the KDE Bugtracking System to requesting Secret Service support on `plasma-nm`. Someone else opened [458318](https://bugs.kde.org/show_bug.cgi?id=458318) calling for the deprecation of `kwalletd5` to encourage this change.

The [issue](https://invent.kde.org/plasma/plasma-nm/-/issues/7) recently made on `plasma-nm`'s issue tracker was closed 2 months ago for being a duplicate of the original bug report, and there doesn't appear to be any progress on it.

Would anyone mind sharing a temporary workaround or how they currently go about storing `NetworkManager` passwords in KPXC?

**Comment by Lgmrszd** [comment]
> Would anyone mind sharing a temporary workaround or how they currently go about storing `NetworkManager` passwords in KPXC?

You could install nm-applet/networkmanagerapplet from gnome and use that. It uses SecretService and as a result, will request passwords from KPXC. The drawbacks are obviously the fact that it is a gnome applet and will not have same look and feel as kde components.

**Comment by hpfr** [comment]
Maybe this could close as not planned now? Unfortunate that some applications like `plasma-nm` still lack support for `org.freedesktop.secrets` for now, and it looks like @michaelk83’s deprecation request in https://github.com/keepassxreboot/keepassxc/issues/3679#issuecomment-1227689257 has gone unanswered, but it seems tough for KeePassXC to justify `org.kde.kwalletd5` support at this point, no? The issue is still a good place to collect workarounds of course.

**Comment by droidmonkey** [comment]
Concur, this won't be implemented in KeePassXC just to appease one application that is still left behind. We offer many other options to provide credentials such as auto-type.

**Comment by michaelk83** [comment]
If anyone is interested, there's an [ongoing MR](https://invent.kde.org/frameworks/kwallet/-/merge_requests/97) to split KWallet daemon into a Secret Service backend and a KWallet API proxy daemon. Once that's merged, the Secret Service backend can be swapped out for other Secret Service implementations, including KeePassXC.

**Comment by rocketraman** [comment]
> If anyone is interested, there's an [ongoing MR](https://invent.kde.org/frameworks/kwallet/-/merge_requests/97) to split KWallet daemon into a Secret Service backend and a KWallet API proxy daemon. Once that's merged, the Secret Service backend can be swapped out for other Secret Service implementations, including KeePassXC.

That was merged and will be available in [Frameworks 6.14](https://blogs.kde.org/2025/04/19/this-week-in-plasma-many-many-things/#frameworks-614-3). Here is a blog post about it: https://notmart.org/blog/2025/04/towards-a-transition-from-kwallet-to-secret-service/, including an experimental facility to migrate data from KWallet into the new secret service backend (which could be KeepassXC).

**Comment by nicheosala** [comment]
> > If anyone is interested, there's an [ongoing MR](https://invent.kde.org/frameworks/kwallet/-/merge_requests/97) to split KWallet daemon into a Secret Service backend and a KWallet API proxy daemon. Once that's merged, the Secret Service backend can be swapped out for other Secret Service implementations, including KeePassXC.
> 
> That was merged and will be available in [Frameworks 6.14](https://blogs.kde.org/2025/04/19/this-week-in-plasma-many-many-things/#frameworks-614-3). Here is a blog post about it: https://notmart.org/blog/2025/04/towards-a-transition-from-kwallet-to-secret-service/, including an experimental facility to migrate data from KWallet into the new secret service backend (which could be KeepassXC).

Wow! Have you tried migrating yet?

**Comment by Lgmrszd** [comment]
> Wow! Have you tried migrating yet?

@nicheosala  I did. I set the parameters as stated in the linked article, and after recent update I got prompted to migrate my kwallet. It is not smooth, first time KeePassXC asked me to create a new DB and I declined, and so the wallet seem to be lost, but after relogin, same thing happened, I created a new KPXC DB just for kde wallet and it worked (in fact, it created one KeePass database per wallet, I had an empty second kde wallet). Now it mostly seem to work, kmail works fine for example, but network manager fails to query wifi passwords.

**Comment by michaelk83** [comment]
> but network manager fails to query wifi passwords.

See https://bugs.kde.org/show_bug.cgi?id=504312 for that.

**Comment by cryptomilk** [comment]
I tried a lot of times to create new DBs and always failed but at one point it just migrated to my existing keepassxc database. However you can also merge keepassxc databases ...

**Comment by rocketraman** [comment]
> Wow! Have you tried migrating yet?

Yes, but like other users it was not smooth. However, I think the following process is what should work quite nicely (please post any corrections, I obviously did not follow this process start to finish as I was messing with things).

1. Create a Group in KeepassXC for future libsecret secrets.
2. Using kwalletmanager did "Close" of my "kdewallet" wallet.
3. In `Settings, Configure Wallet`, changed "Use KWallet for the Secret Service interface" to disabled.
4. Kill the `ksecretd` process (I assumed step 3 would do this automatically but it did not).
5. Turned on Secret Service Integration in keepassxc and selected the Group I created in Step 1 under "Expose entries under this group".
6. Edited `~/.config/kwalletrc` to include the required entries:
    ```
    [KSecretD]
    Enabled=false
    
    [Migration]
    MigrateTo3rdParty=true
    ```
7. Kill the `kwalletd6` process (I assumed "Close" of the wallet in step 2 would do this automatically but it did not).
8. KDE should restart kwalletd6 and it will execute the migration -- the entries will populate into the correct group in keepassxc immediately.
9. In kwalletmanager, your keepassxc database should now show up as a wallet. Open it.

All done. Writing entries in kwallet should update keepassxc.

**Comment by michaelk83** [comment]
Your step 3 is implied by steps 4 and 6. "Use KWallet for the Secret Service interface" makes no difference if `ksecretd` is not running. This setting also isn't applied until `ksecretd` is restarted. I've opened [kde/504693](https://bugs.kde.org/show_bug.cgi?id=504693) to address that.
Step 2 is implied by steps 4 and 7.

So this can be rearranged as follows (not tested!):
1. Make sure you don't have any unwanted Secret Service providers running or DBus activatable, such as Gnome Keyring. Uninstall them if necessary.
1. Edit ~/.config/kwalletrc to include the required entries: 
```
[KSecretD]
Enabled=false

[Migration]
MigrateTo3rdParty=true
```
3. Kill `ksecretd` and `kwalletd6`.
    - If you still have an unwanted Secret Service provider at this point, it may take over org.freedesktop.secrets, and KWallet may export your entries there. Which is why step 1 is recommended first!
1. Create a Group in KeepassXC for Secret Service, expose it to Secret Service, then enable Secret Service Integration.
    - I recommend using a separate DB for Secret Service.
1. Start `kwalletd6` again, e.g. by launching KWalletManager. This should initiate the migration.

@Lgmrszd @cryptomilk @rocketraman What were your sticking points when migrating? Was it just the DB creation and enabling integration, or were there other issues?

**Comment by rocketraman** [comment]
> Your step 3 is implied by steps 4 and 6. "Use KWallet for the Secret Service interface" makes no difference if `ksecretd` is not running. This setting also isn't applied until `ksecretd` is restarted. I've opened [kde/504693](https://bugs.kde.org/show_bug.cgi?id=504693) to address that. Step 2 is implied by steps 4 and 7.

Ok. Note that I expected the GUI setting in Step 3 to set this in the config file:

```
[KSecretD]
Enabled=false
```

but instead it set:

```
[org.freedesktop.secrets]
apiEnabled=false
```

which was a bit confusing.

> 
> So this can be rearranged as follows (not tested!):
> 
> 1. Make sure you don't have any unwanted Secret Service providers running or DBus activatable, such as Gnome Keyring. Uninstall them if necessary.
> 2. Edit ~/.config/kwalletrc to include the required entries:
> 
> ```
> [KSecretD]
> Enabled=false
> 
> [Migration]
> MigrateTo3rdParty=true
> ```
> 
> 3. Kill `ksecretd` and `kwalletd6`.
>    
>    * If you still have an unwanted Secret Service provider at this point, it may take over org.freedesktop.secrets, and KWallet may export your entries there. Which is why step 1 is recommended first!
> 4. Create a Group in KeepassXC for Secret Service, expose it to Secret Service, then enable Secret Service Integration.
>    
>    * I recommend using a separate DB for Secret Service.
> 5. Start `kwalletd6` again, e.g. by launching KWalletManager. This should initiate the migration.

While I think it will work fine, this approach feels less safe. When kwalletd6 is killed at step 3, the system will likely automatically restart it through dbus activation or systemd (not sure which). At this point the migration will begin. Presumably the migration waits for a secret service to exist so it will still work as you laid it out, but doing ksecretd and enabling keepassxc first ensures that a) keepassxc is the only libsecret running (otherwise it will refuse to be enabled), and b) it is definitely configured correctly before kwalletd restarts and migrates.

> I recommend using a separate DB for Secret Service.

Why? Having everything inside one DB makes backups easier.

> [@Lgmrszd](https://github.com/Lgmrszd) [@cryptomilk](https://github.com/cryptomilk) [@rocketraman](https://github.com/rocketraman) What were your sticking points when migrating? Was it just the DB creation and enabling integration, or were there other issues?

I tried it with krdc, Plasma Network Manager opening a wireguard VPN connection, and Jetbrains toolbox (which appears to use the kwalletd6 API). All retrieved the information from keepassxc without problem.

Then all of a sudden everything stopped working. I finally realized that gnome-keyring-daemon (which I didn't even know I had installed) decided to make an appearance and somehow enabled itself, without keepassxc reporting any issues or conflicting secret services. All my logins stopped working until I disabled the secret service in keepassxc in order to debug the issue, which only *then* reported that another secret service was running. Aha. It would be helpful if the Keepass warning about other secret services was *always* shown, regardless of whether it was already enabled or not. And I'm not sure why gnome-keyring-daemon never enabled itself when kwalletd was the secret service, but did when keepassxc was.

**Comment by rocketraman** [comment]
Another question I have is: does the keepassxc "wallet" have to be "Open" in kwalletmanager for apps using the kwallet API to work? It doesn't appear to be needed -- logins appear to work with it Closed as well as Open. But some clarity on this would be useful.

**Comment by michaelk83** [comment]
> Note that I expected the GUI setting in Step 3 to set ... which was a bit confusing.

Yea, those are separate settings. "Use KWallet for the Secret Service interface" still allows KSecretD to run. It just tells it to not expose the Secret Service API on the public DBus address. It's still used internally by KWallet via a "private" DBus address.

OTOH, `[KSecretD] Enabled=false` completely disables KSecretD, which is required to migrate to a 3rd party Secret Service (and to use it via the KWallet API proxy). This setting currently isn't exposed in the KWallet GUI.

> > So this can be rearranged as follows (not tested!):
> >    ...
> 
> While I think it will work fine, this approach feels less safe. When kwalletd6 is killed at step 3, the system will likely automatically restart it through dbus activation or systemd (not sure which). At this point the migration will begin.

This is why I added the extra step 1 to make sure you don't have any unwanted Secret Service providers. If there are no other providers, there's nowhere else to migrate to. I've also just added [kde/504709](https://bugs.kde.org/show_bug.cgi?id=504709) to implement more control over the migration.

> > I recommend using a separate DB for Secret Service.
> 
> Why? Having everything inside one DB makes backups easier.

- The Secret Service entries are always local to each machine, whereas the other entries usually are not (e.g. can be shared to a mobile device or another PC).
- The Secret Service entries are generated by the client apps, and normally shouldn't be touched directly by the end user. The other entries are user-controlled.
- And last but not least, some or all of the other entries may be more sensitive than the local app passwords, and may warrant tighter security. For example, you may want to keep the Secret Service DB unlocked for longer and more often than some of your other passwords.

> gnome-keyring-daemon (which I didn't even know I had installed) decided to make an appearance

Gnome Keyring is a known annoyance. I agree that KeePassXC's warning could be handled better.

> Another question I have is: does the keepassxc "wallet" have to be "Open" in kwalletmanager for apps using the kwallet API to work? It doesn't appear to be needed

The collection must be unlocked in KeePassXC, which *should* in turn reflect as "open" in KWalletManager. But I've previously [reported a bug](https://bugs.kde.org/show_bug.cgi?id=459287) where this doesn't always work correctly. This is probably only a display issue.

**Comment by rocketraman** [comment]
@michaelk83  One other issue I've seen with the integration is a bunch of "almost empty" entries are created in KeepassXC. They have a username but nothing else. No title, password, or any attributes at all. I also see nothing in KWallet Manager that would correspond to these entries.

My next step is to delete these and see if they come back.

**Comment by michaelk83** [comment]
Can you give some examples of such usernames?

**Comment by rocketraman** [comment]
> Can you give some examples of such usernames?

They are all `rocketraman` like my GitHub username. I don't use this for local services but it's a common username I use for online services.

**Comment by michaelk83** [comment]
Hmm.. you'll have to figure out on your own where those came from. It doesn't sound like it came from the migration (though who knows....)

I did run into a bug in KWalletManager where it failed to show me any items that I created via Secret Service, so that might be why you're not seeing those there. You can inspect things more reliably with DBus tools such as `qdbus` or in Seahorse (aside from KeePassXC). Though Seahorse only uses the Secret Service API, not the KWallet API.

**Comment by rocketraman** [comment]
> It doesn't sound like it came from the migration (though who knows....)

I believe they were created by something after the migration.

> Hmm.. you'll have to figure out on your own where those came from.

Hard to do that when the entries have no origin information. It would be nice if KeepassXC stored some origin information in the entry attributes.

> You can inspect things more reliably with DBus tools such as `qdbus` or in Seahorse (aside from KeePassXC). Though Seahorse only uses the Secret Service API, not the KWallet API.

Seahorse just shows them as blank entries as well.

I deleted them and so far they have not been created again. But we'll see.

At the end of the day I don't think this is KeepassXC related (except it would be nice if KeepassXC stored some process origin information), but probably more of an errant use of the libsecret API by some process on my system.

**Comment by michaelk83** [comment]
> I did run into a bug in KWalletManager where it failed to show me any items that I created via Secret Service

Turns out it only supports the QtKeyChain schema. Anything else will be ignored by KWalletManager. Items with no attributes obviously don't fit that schema.

**Comment by dougg0k** [comment]
I noticed that the first time I use an app that requests keepassxc to open as the wallet. It opens normally and requests for the permission and it works.

However if I keep keepassxc open, and close the app that requests it and try again. The next time it tries to create a database. If I close the prompt, it complains as if there were no OS keyring available. 

If I close both, and try to open the app that requests it, it doesnt retry to reopen keepassxc, it straight complains that thre is no OS keyring available.

Also, if I say to remember the request that was allowed now and future, it always asking regardless, every time. In the first time that is.

So, somewhat broken. Not viable to be used the way it is.

```
[KSecretD]
Enabled=false

[Migration]
MigrateTo3rdParty=true
WalletsMigratedToSecretService=kdewallet

[Wallet]
Close When Idle=false
Close on Screensaver=false
Default Wallet=some_kdbx_wallet
Enabled=true
First Use=false
Idle Timeout=10
Launch Manager=false
Leave Manager Open=false
Leave Open=false
Prompt on Open=false
Use One Wallet=true

[org.freedesktop.secrets]
apiEnabled=true
```

Anyone know any fix to this?

**Comment by rocketraman** [comment]
@dougg0k I just ran into this (or something similar) also. It seems that somehow the KeepassXC secret service was stopped or corrupted. The workaround was to disable and re-enable the KeepassXC secret service i.e. Tools, Settings, disable, hit OK, and Tools, Settings and enable it again.

**Comment by dougg0k** [comment]
Nah, definitely not the case on that.

---

Something that may work better on chromium based apps. `--password-store="gnome-libsecret"` Regardless if on KDE or not.

**Comment by michaelk83** [comment]
@dougg0k There are a few separate issues here:

> if I keep keepassxc open, and close the app that requests it and try again. The next time it tries to create a database.

Sounds like something got messed up on KeePassXC side. The disable and re-enable workaround mentioned [above](#issuecomment-2977584239) may work for this case.

> If I close the prompt, it complains as if there were no OS keyring available.

If it's asking you to create a DB, it thinks there is no DB assigned to Secret Service. If you cancel that prompt, then it still thinks there is no relevant DB.
It may also ask you to unlock the DB if it's locked.

> If I close both, and try to open the app that requests it, it doesnt retry to reopen keepassxc, it straight complains that thre is no OS keyring available.

You may be missing a DBus auto-start file for Secret Service, or it's not pointing at KeePassXC.
See https://wiki.archlinux.org/title/KDE_Wallet#Automatic_D-Bus_activation for such auto-start file example with KWallet.

> Also, if I say to remember the request that was allowed now and future, it always asking regardless, every time. In the first time that is.

It's a known limitation at the moment: #7571 
Authorization is only remembered for as long as both the client app and the KeePassXC processes are running.

**Comment by dougg0k** [comment]
The issue was somewhat fixed after setting the store.

At least in my opinion, even without those issues, keepassxc is not viable to be used as secret service / wallet alternative, maybe it would be once they fix the always asking to to allow the request, even if you set to remember.

Another one is missing FIDO2 support that would make possible to use passkey instead of password, making the experience better / not as bad.

I stopped using keepassxc for it.

**Comment by hueychen27** [comment]
`systemctl --user restart dbus-:1.1-org.kde.kwalletd6@0.service` has to be run every time a KeepassXC database is reopened to use secrets for kwallet6.
